File: app/src/main/java/org/ole/planet/myplanet/ui/viewer/PDFReaderActivity.java
package org.ole.planet.myplanet.ui.viewer;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.github.barteksc.pdfviewer.listener.OnLoadCompleteListener;
import com.github.barteksc.pdfviewer.listener.OnPageChangeListener;
import com.github.barteksc.pdfviewer.listener.OnPageErrorListener;
import com.github.barteksc.pdfviewer.scroll.DefaultScrollHandle;

import org.ole.planet.myplanet.R;
import org.ole.planet.myplanet.databinding.ActivityPdfreaderBinding;
import org.ole.planet.myplanet.datamanager.DatabaseService;
import org.ole.planet.myplanet.model.RealmMyLibrary;
import org.ole.planet.myplanet.service.AudioRecorderService;
import org.ole.planet.myplanet.ui.library.AddResourceFragment;
import org.ole.planet.myplanet.utilities.IntentUtils;
import org.ole.planet.myplanet.utilities.NotificationUtil;
import org.ole.planet.myplanet.utilities.Utilities;

import java.io.File;

import io.realm.Realm;

public class PDFReaderActivity extends AppCompatActivity implements OnPageChangeListener, OnLoadCompleteListener, OnPageErrorListener, AudioRecorderService.AudioRecordListener {
    private ActivityPdfreaderBinding activityPdfreaderBinding;
    private static final String TAG = "PDF Reader Log";
    private String fileName;
    private AudioRecorderService audioRecorderService;
    private RealmMyLibrary library;
    private Realm mRealm;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activityPdfreaderBinding = ActivityPdfreaderBinding.inflate(getLayoutInflater());
        setContentView(activityPdfreaderBinding.getRoot());
        audioRecorderService = new AudioRecorderService().setAudioRecordListener(this);
        mRealm = new DatabaseService(this).getRealmInstance();
        if (getIntent().hasExtra("resourceId")) {
            String resourceID = getIntent().getStringExtra("resourceId");
            library = mRealm.where(RealmMyLibrary.class).equalTo("id", resourceID).findFirst();
        }
        renderPdfFile();

        activityPdfreaderBinding.fabRecord.setOnClickListener(view -> {
            if (audioRecorderService.isRecording()) {
                audioRecorderService.stopRecording();
            } else {
                audioRecorderService.startRecording();
            }
        });

        activityPdfreaderBinding.fabPlay.setOnClickListener(view -> {
            if (library != null && !TextUtils.isEmpty(library.translationAudioPath)) {
                IntentUtils.openAudioFile(this, library.translationAudioPath);
            }
        });
    }

    private void renderPdfFile() {
        Intent pdfOpenIntent = getIntent();
        fileName = pdfOpenIntent.getStringExtra("TOUCHED_FILE");
        if (fileName != null && !fileName.isEmpty()) {
            activityPdfreaderBinding.pdfFileName.setText(fileName);
            activityPdfreaderBinding.pdfFileName.setVisibility(View.VISIBLE);
        }

        File file = new File(getExternalFilesDir(null), "ole/" + fileName);
        if (file.exists()) {
            try {
                Utilities.log(file.getAbsolutePath());
                activityPdfreaderBinding.pdfView.fromFile(file).defaultPage(0).enableAnnotationRendering(true).onLoad(this).onPageChange(this).scrollHandle(new DefaultScrollHandle(this)).load();
            } catch (Exception e) {
                e.printStackTrace();
                Toast.makeText(getApplicationContext(), getString(R.string.unable_to_load) + fileName, Toast.LENGTH_LONG).show();
            }
        } else {
            Toast.makeText(getApplicationContext(), "File not found: " + fileName, Toast.LENGTH_LONG).show();
        }
    }

    @Override
    public void loadComplete(int nbPages) {}

    @Override
    public void onPageChanged(int page, int pageCount) {}

    @Override
    public void onPageError(int page, Throwable t) {
        Log.e(TAG, "Cannot load page " + page);
    }

    @Override
    public void onRecordStarted() {
        Utilities.toast(this, getString(R.string.recording_started));
        NotificationUtil.create(this, R.drawable.ic_mic, "Recording Audio", getString(R.string.ole_is_recording_audio));
        activityPdfreaderBinding.fabRecord.setImageResource(R.drawable.ic_stop);
    }

    @Override
    public void onRecordStopped(String outputFile) {
        Utilities.toast(this, getString(R.string.recording_stopped));
        NotificationUtil.cancellAll(this);
        if (outputFile != null) {
            updateTranslation(outputFile);
            AddResourceFragment.showAlert(this, outputFile);
        }
        activityPdfreaderBinding.fabRecord.setImageResource(R.drawable.ic_mic);
    }

    private void updateTranslation(String outputFile) {
        if (library != null) {
            if (!mRealm.isInTransaction()) mRealm.beginTransaction();
            library.translationAudioPath = outputFile;
            mRealm.commitTransaction();
            Utilities.toast(this, getString(R.string.audio_file_saved_in_database));
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (audioRecorderService != null && audioRecorderService.isRecording())
            audioRecorderService.stopRecording();
    }

    @Override
    public void onError(String error) {
        NotificationUtil.cancellAll(this);
        Utilities.toast(this, error);
        activityPdfreaderBinding.fabRecord.setImageResource(R.drawable.ic_mic);
    }
}

File: app/src/main/java/org/ole/planet/myplanet/ui/viewer/PDFReaderActivity.kt
package org.ole.planet.myplanet.ui.viewer

import android.graphics.pdf.PdfRenderer
import android.os.Bundle
import android.os.ParcelFileDescriptor
import android.text.TextUtils
import android.view.View
import android.widget.ImageView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.graphics.createBitmap
import androidx.lifecycle.lifecycleScope
import dagger.hilt.android.AndroidEntryPoint
import java.io.File
import javax.inject.Inject
import kotlinx.coroutines.launch
import org.ole.planet.myplanet.R
import org.ole.planet.myplanet.databinding.ActivityPdfreaderBinding
import org.ole.planet.myplanet.model.RealmMyLibrary
import org.ole.planet.myplanet.repository.LibraryRepository
import org.ole.planet.myplanet.repository.MyPersonalRepository
import org.ole.planet.myplanet.service.AudioRecorderService
import org.ole.planet.myplanet.service.AudioRecorderService.AudioRecordListener
import org.ole.planet.myplanet.service.UserProfileDbHandler
import org.ole.planet.myplanet.ui.resources.AddResourceFragment
import org.ole.planet.myplanet.utilities.EdgeToEdgeUtils
import org.ole.planet.myplanet.utilities.FileUtils
import org.ole.planet.myplanet.utilities.IntentUtils.openAudioFile
import org.ole.planet.myplanet.utilities.NotificationUtils.cancelAll
import org.ole.planet.myplanet.utilities.NotificationUtils.create
import org.ole.planet.myplanet.utilities.Utilities

@AndroidEntryPoint
class PDFReaderActivity : AppCompatActivity(), AudioRecordListener {
    private lateinit var binding: ActivityPdfreaderBinding
    private lateinit var audioRecorderService: AudioRecorderService
    private var fileName: String? = null
    @Inject
    lateinit var myPersonalRepository: MyPersonalRepository
    @Inject
    lateinit var libraryRepository: LibraryRepository
    @Inject
    lateinit var userProfileDbHandler: UserProfileDbHandler
    private lateinit var library: RealmMyLibrary
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityPdfreaderBinding.inflate(layoutInflater)
        setContentView(binding.root)
        EdgeToEdgeUtils.setupEdgeToEdge(this, binding.root)
        audioRecorderService = AudioRecorderService().setAudioRecordListener(this)
        audioRecorderService.setCaller(this, this)
        if (intent.hasExtra("resourceId")) {
            val resourceID = intent.getStringExtra("resourceId")
            lifecycleScope.launch {
                resourceID?.let {
                    library = libraryRepository.getLibraryItemById(it)!!
                }
            }
        }
        renderPdfFile()
        binding.fabRecord.setOnClickListener { audioRecorderService.onRecordClicked()}
        binding.fabPlay.setOnClickListener {
            if (this::library.isInitialized && !TextUtils.isEmpty(library.translationAudioPath)) {
                openAudioFile(this, library.translationAudioPath)
            }
        }
    }

    private fun renderPdfFile() {
        val pdfOpenIntent = intent
        fileName = pdfOpenIntent.getStringExtra("TOUCHED_FILE")
        if (!fileName.isNullOrEmpty()) {
            binding.pdfFileName.text = FileUtils.nameWithoutExtension(fileName)
            binding.pdfFileName.visibility = View.VISIBLE
        } else {
            binding.pdfFileName.text = getString(R.string.message_placeholder, "No file selected")
            binding.pdfFileName.visibility = View.VISIBLE
        }
        val file = File(getExternalFilesDir(null), "ole/$fileName")
        if (file.exists()) {
            try {
                val fileDescriptor = ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY)
                val pdfRenderer = PdfRenderer(fileDescriptor)
                val page = pdfRenderer.openPage(0)
                val bitmap = createBitmap(page.width, page.height)
                page.render(bitmap, null, null, PdfRenderer.Page.RENDER_MODE_FOR_DISPLAY)

                val imageView = ImageView(this)
                imageView.setImageBitmap(bitmap)
                imageView.scaleType = ImageView.ScaleType.FIT_CENTER

                binding.pdfPlaceholder.visibility = View.GONE
                val parent = binding.pdfPlaceholder.parent as android.view.ViewGroup
                parent.addView(imageView, android.view.ViewGroup.LayoutParams(android.view.ViewGroup.LayoutParams.MATCH_PARENT, 0))
                (imageView.layoutParams as android.widget.LinearLayout.LayoutParams).weight = 1f

                page.close()
                pdfRenderer.close()
                fileDescriptor.close()
            } catch (e: Exception) {
                e.printStackTrace()
                Toast.makeText(applicationContext, getString(R.string.unable_to_load) + fileName, Toast.LENGTH_LONG).show()
            }
        } else {
            Toast.makeText(applicationContext, "File not found: $fileName", Toast.LENGTH_LONG)
                .show()
        }
    }


    override fun onRecordStarted() {
        Utilities.toast(this, getString(R.string.recording_started))
        create(this, R.drawable.ic_mic, "Recording Audio", getString(R.string.ole_is_recording_audio))
        binding.fabRecord.setImageResource(R.drawable.ic_stop)
    }

    override fun onRecordStopped(outputFile: String?) {
        Utilities.toast(this, getString(R.string.recording_stopped))
        cancelAll(this)
        updateTranslation(outputFile)
        val userModel = userProfileDbHandler.userModel
        if (userModel != null) {
            AddResourceFragment.showAlert(
                this,
                outputFile,
                myPersonalRepository,
                userModel.id,
                userModel._id,
                userModel.name
            )
        }
        binding.fabRecord.setImageResource(R.drawable.ic_mic)
    }

    private fun updateTranslation(outputFile: String?) {
        if (this::library.isInitialized) {
            lifecycleScope.launch {
                libraryRepository.updateLibraryItem(library.id!!) {
                    it.translationAudioPath = outputFile
                }
                library.translationAudioPath = outputFile
                Utilities.toast(
                    this@PDFReaderActivity,
                    getString(R.string.audio_file_saved_in_database)
                )
            }
        }
    }

    override fun onDestroy() {
        if (this::audioRecorderService.isInitialized && audioRecorderService.isRecording()) {
            audioRecorderService.stopRecording()
        }
        userProfileDbHandler.onDestroy()
        super.onDestroy()
    }

    override fun onError(error: String?) {
        cancelAll(this)
        Utilities.toast(this, error)
        binding.fabRecord.setImageResource(R.drawable.ic_mic)
    }
}
